core/src/main/scala/kafka/api/FetchResponse.scala

```diff
@@ -48,6 +49,36 @@ object FetchResponsePartitionData {
 
 case class FetchResponsePartitionData(error: Short = Errors.NONE.code, hw: Long = -1L, messages: MessageSet) {
   val sizeInBytes = FetchResponsePartitionData.headerSize + messages.sizeInBytes
+
+  def toMessageFormat(toMagicValue: Byte): FetchResponsePartitionData = {
```

Comment:
I agree it makes sense to put the message set format conversion code block into MessageSet instead of here.

However, I make the IO threads to do the conversion on purpose because I want to share the workload between KafkaApis threads and IO threads. Typically, IO threads are more light-weighted than KafkaApis threads. If we have to make conversion for some requests, I am trying to put the conversion load on IO threads.